jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/resource/ResourceFactory.java

```diff
@@ -15,29 +15,298 @@
+     *
+     * @param resource the relative name of the resource
+     * @return Resource or null
+     */
+    default Resource newClassPathResource(String resource)
+    {
+        URL url = Resource.class.getResource(resource);
+
+        if (url == null)
+            url = Loader.getResource(resource);
+        if (url == null)
+            return null;
+        try
+        {
+            URI uri = url.toURI();
+            return Resource.create(uri);
+        }
+        catch (URISyntaxException e)
+        {
+            throw new IllegalArgumentException(e);
+        }
+    }
+
+    default Resource newResource(String resource)
+    {
+        return newResource(Resource.toURI(resource));
+    }
+
+    default Resource newResource(Path path)
+    {
+        return newResource(path.toUri());
+    }
+
+    default ResourceCollection newResource(Collection<URI> uris)
+    {
+        return Resource.combine(uris.stream().map(this::newResource).toList());
+    }
+
+    default Resource newResource(URL url)
+    {
+        try
+        {
+            return newResource(url.toURI());
+        }
+        catch (URISyntaxException e)
+        {
+            throw new IllegalArgumentException("Error creating resource from URL: " + url, e);
+        }
+    }
+
+    default Resource newJarFileResource(URI uri)
+    {
+        if (!FileID.isArchive(uri))
+            throw new IllegalArgumentException("Path is not a Java Archive: " + uri);
+        if (!uri.getScheme().equalsIgnoreCase("file"))
+            throw new IllegalArgumentException("Not an allowed path: " + uri);
+        return newResource(URIUtil.toJarFileUri(uri));
+    }
+
+    static ResourceFactory root()
+    {
+        return __ROOT;
+    }
+
+    static ResourceFactory.Closeable closeable()
+    {
+        return new Closeable();
+    }
+
+    static ResourceFactory.LifeCycle lifecycle()
+    {
+        LifeCycle factory = new LifeCycle();
+        org.eclipse.jetty.util.component.LifeCycle.start(factory);
+        return factory;
+    }
+
+    static ResourceFactory of(Resource baseResource)
+    {
+        Objects.requireNonNull(baseResource);
+
+        if (baseResource instanceof ResourceFactory resourceFactory)
+            return resourceFactory;
+
+        return new ResourceFactory()
+        {
+            @Override
+            public Resource newResource(URI resource)
+            {
+                // TODO add an optimized pathway that keeps the URI and doesn't go via String
+                return newResource(resource.toString());
+            }
+
+            @Override
+            public Resource newResource(String resource)
+            {
+                return baseResource.resolve(resource);
+            }
+        };
+    }
+
+    static ResourceFactory of(Container container)
+    {
+        Objects.requireNonNull(container);
+
+        LifeCycle factory = container.getBean(LifeCycle.class);
```

Comment:
Worry about thread safe execution?
What if this `of(Container)` is called from multiple threads?
Something like `container.calculateIfAbsent`?

Reply:
might just make sense to synchronize here (on the container?)

Reply:
Excellent catch!

This is not easy as it looks as there is nothing to rely on that offers an atomic `compute` semantic, and using a lock here might be overkill.

I'll leave this as-is for now, with a TODO.